src/engine/builtins.c

```diff
@@ -2515,15 +2506,16 @@ LIST * builtin_shell( FRAME * frame, int flags )
         buffer[ ret ] = 0;
         if ( !no_output_opt )
         {
-            if ( strip_eol_opt )
-                rtrim( buffer );
             string_append( &s, buffer );
         }
 
         /* Explicit EOF check for systems with broken fread */
         if ( feof( p ) ) break;
     }
 
+    if ( strip_eol_opt )
+        string_rtrim( &s );
```

Comment:
By moving it outside the loop and only processing the final string, the bug is mitigated.